ykrt/src/compile/j2/opt/known_bits.rs

```diff
@@ -0,0 +1,482 @@
+//! Known bits analysis.
+//!
+//! Data-flow analysis to gain information about bits.
+//! We use similar ideas to PyPy:
+//! https://pypy.org/posts/2024/08/toy-knownbits.html
+//!
+
+use crate::compile::{
+    j2::{
+        hir::{And, Const, ConstKind, Inst, InstIdx, InstT, Or, Ty},
+        opt::{
+            BlockLikeT,
+            fullopt::{CommitInstOpt, OptOutcome, PassOpt, PassT},
+        },
+    },
+    jitc_yk::arbbitint::ArbBitInt,
+};
+use index_vec::IndexVec;
+
+/// Known-bits analysis.
+pub(super) struct KnownBits {
+    known_bits: IndexVec<InstIdx, Option<KnownBitValue>>,
+    pending_commit: Option<KnownBitValue>,
```

Comment:
This attribute needs documenting (I know what it does because I know its existence is because of an incomplete API... which is shameful on my part!).